### ISSUE TITLE
Continue with anonymous block arguments

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#8764]
Follows up on #8117. This PR added syntax for anonymous block arguments `def foo(&)`. The idea is to use this syntax exclusively for yielding methods.
Currently it's valid to completely omit the block argument when there's `yield` in the body. This hides the yielding nature of the method from the signature and API docs. Such an essential property should be made visible.

After the initial syntax was added, we switched usage in stdlib in #8394. 

The next step would be to deprecate some of the currently valid method signatures. The idea as proposed in #8117:

- Just `&` means a non-captured block argument: because it doesn't have a name there's no way to capture it
- `&block` means a captured block argument, and the compiler will verify that it's used in the semantic phase (because it might be mentioned inside a macro)
- `&block` without a mention will give a compile error suggesting to either change it to `&` to avoid a closure or to mention it somehow
- if you `yield` and the method signature doesn't have `&` the compiler will tell you

This would be a breaking change for yielding methods with a named block argument or no block argument at all, and also for methods with a named block argument that's never used in the literal scope of its body. But it improves clarity and helps to better identify the behaviour of block arguments.

**Comment by RX14** [comment]
> This hides the yielding nature of the method from the signature and API docs.

No, it shows up in the API docs if you use `yield`. We shouldn't force people to annotate their method with a useless `&` if they use `yield`. We should make a distinction between `&` never capturing and `&block` always capturing though.

**Comment by straight-shoota** [comment]
Oh my bad. The API docs already represent this correctly.

I'd still prefer always specifying `&` on yielding methods (types can be omitted obviously). Even if it's technically unnecessary, it still improves readability of the source code. We already put `&` in the signature in the API docs, but reading code is almost as important.
I acknowledge that this doesn't need to be enforced by the compiler, though. It could just be a style guide.

**Comment by HertzDevil** [comment]
It seems [Ruby eventually wants to require `&` in methods that accept a block too](https://bugs.ruby-lang.org/issues/11256?tab=history#note-16).

For context, Ruby 3.1 allows a bare `&` to forward the anonymous block argument, whether it is captured or not:

```ruby
def foo1(&)
  yield
end

def bar1(&)
  foo1(&)
end

bar1 { 1 } # => 1

def foo2(&block)
  block.call
end

def bar2(&)
  proc(&) # captures anonymous block
  foo2(&)
end

bar2 { 1 } # => 1
```

It is not a syntactic transformation, from what I can tell.

**Comment by zw963** [comment]
> No, it shows up in the API docs if you use `yield`. We shouldn't force people to annotate their method with an unneccesary `&` if they use `yield` but don't want to annotate the arg types. The rest of the changes are great.

i don't agree with this idea, if yield exists in method body, block must be provided when invoke this method, add a extra & in method parameter is more clear way reveal this.

Also see following code for a really use case need to improve in std-lib.

https://github.com/crystal-lang/crystal/blob/771bddce1b39830316f03136a1cf8245aa976523/src/socket/server.cr#L48-L55

```crystal
```

See [discuss](https://forum.crystal-lang.org/t/why-unixserver-only-override-accept-method-but-leave-accept-not-redefined/5167/5?u=zw963) here, A extra & in the pamameter will make issue like this more clearly, user don't need search `yield` in (probalby a big) method body too find out whether this method need a block.
